diag_manager/fms_diag_object_container.F90

```diff
@@ -202,10 +202,20 @@ end function diag_obj_iterator_constructor
    function diag_object_container_constructor () result (doc)
       type(FmsDiagObjectContainer_t), allocatable :: doc !< The resultant container.
       allocate(doc)
-      doc%the_linked_list = FmsDlList_t()
-      !! print * , "In DOC constructor"
+      allocate(doc%the_linked_list)
```

Comment:
Does the destructor get called every time?  I think you were getting a seg fault because you weren't deallocating this variable.  If that's the case and you are allocating a pointer that's been allocated, then you are leaking memory.

Reply:
Can this be changed to 
```fortran
doc%the_linked_list => null
```
Are you setting this equal to something, or using it as a pointer?

Reply:
Before and now linked list class (and in the container class) , have a function decorated with the “final” keyword.  When class instance goes out of scope, the final decorated function cleans up any links (nodes) that remain.

Before,  linked list class had an OO-stype constructor (function with exactly the same name as the type) used in the unit test like this :

ll = FmsLlList_t()

What I coul see as I stepped though it in the debugger for Intel compiled versions, was that as soon as the RHS above finished executing, deallocation of anything that was “allocatable” in the linked list was automatically happening.  So for one instance created by constructor related functions, members head and tail were manually allocated, automatically deallocated when the instance went out of scope, and finally, with the “=” above, memberwise copied to ll.

In this version I got rid of the FmsLlList_t() constructor function. The final keywork is important, and the worked differently Intel vs gcc, so not a great loss. Now must explicitly do :
ll = allocate(FmsllList_t)
call ll%initialize()

For having the type FmsLlList_t() clean up after itself, users can just have its instance be allocatable.
In this case the final keyword decoration on function “destructor()” will make sure function 
clear() (or clear_all()) is called and all pointers data will be deallocated. If the user makes the class instance a pointer, then the user is responsible for calling function clear() when desired or before end of program execution.

Reply:
Concerning your second question, I think yes. Do you mean change lines 204  - 205 from 
```
allocate(doc)
allocate(doc%the_linked_list)
```
to 
```
doc=>null()
allocate(doc)
allocate(doc%the_linked_list)
```

Reply:
or
```
doc=>null()
allocate(doc)
doc%the_linked_list => null()
allocate(doc%the_linked_list)
```